tfjs-layers/src/layers/nlp/tokenizers.ts

```diff
@@ -0,0 +1,134 @@
+/**
+ * @license
+ * Copyright 2023 Google LLC.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * =============================================================================
+ */
+
+/**
+ *  Tokenizer layers.
+ */
+
+/* Original source: keras-nlp/tokenizer.py */
+import { Tensor1D, serialization, tensor1d } from '@tensorflow/tfjs-core';
+
+import { Layer } from '../../engine/topology';
+import { NotImplementedError, ValueError } from '../../errors';
+
+export declare interface TokenizerOptions {
+  mode?: 'tokenize' | 'detokenize'
+}
+
+/**
+ * Base class for Tokenizers.
+ *
+ * Subclassers should always implement the `tokenize()` method, which will also
+ * be the default when calling the layer directly on inputs.
+ */
+export abstract class Tokenizer extends Layer {
+  /**
+   * Transform input tensors of strings into output tokens.
+   *
+   * @param inputs Input tensor.
+   * @param kwargs Additional keyword arguments.
+   */
+  abstract tokenize(inputs: Tensor1D): Tensor1D[];
+
+  /**
+   * Transform tokens back into strings.
+   *
+   * @param inputs Input tensor.
+   * @param kwargs Additional keyword arguments.
+   */
+  detokenize(inputs: Tensor1D[]): Tensor1D {
+    throw new NotImplementedError(
+      `No implementation of 'detokenize()' was found for
+      ${this.constructor.name}.`
+    );
+  }
+
+  /**
+   * Get the tokenizer vocabulary as a list of strings terms.
+   */
+  get vocabulary(): string[] {
+    throw new NotImplementedError(
+      `No implementation of 'vocabulary()' was found for
+      ${this.constructor.name}.`
+    );
+  }
+
+  /**
+   * Returns the total size of the token id space.
+   */
+  get vocabularySize(): number {
+    throw new NotImplementedError(
+      `No implementation of 'vocabularySize()' was found for
+      ${this.constructor.name}.`
+    );
+  }
+
+  /**
+   * Convert an integer id to a string token.
+   */
+  idToToken(id: number): string {
+    throw new NotImplementedError(
+      `No implementation of 'idToToken()' was found for
+      ${this.constructor.name}.`
+    );
+  }
+
+  /**
+   * Convert an integer id to a string token.
+   */
+  tokenToId(token: string): number {
+    throw new NotImplementedError(
+      `No implementation of 'tokenToId()' was found for
+      ${this.constructor.name}.`
+    );
+  }
+
+  override call(inputs: Tensor1D|Tensor1D[], kwargs: TokenizerOptions={mode: 'tokenize'}): Tensor1D|Tensor1D[] {
```

Comment:
Python lets me pass just an array of strings or a string all on its own without converting them to tensors. Ping told me we have some utilities for dealing with these kinds of inputs, but they're not exported from TFJS yet. We could export them and use them in layers to make consuming inputs like these easier (but that can be a separate PR).
* [convertToTensor](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/tensor_util_env.ts#L98) is similar to [tf.convert_to_tensor](https://www.tensorflow.org/api_docs/python/tf/convert_to_tensor), which [BytePairTokenizer uses](https://github.com/keras-team/keras-nlp/blob/master/keras_nlp/tokenizers/byte_pair_tokenizer.py#L500)
* [TensorLike (and similar types)](https://github.com/tensorflow/tfjs/blob/master/tfjs-core/src/types.ts#L141-L164) are types that aren't tensors, but which can be converted to tensors with convertToTensor.

I've also found some interesting behavior in the Keras version with tensors that have more than one dimension. If I use a 2d tensor, it flattens the inner dimensions by concatenating the strings, it seems. This is likely BPE implementation details.
```python
>>> tokenizer.tokenize(tf.constant([['butterfly', 'flybutterfly'], ['flyfly', 'butterbutter']]))
<tf.RaggedTensor [[1, 2, 2, 1, 2], [2, 2, 1, 1]]>
>>> tokenizer.tokenize(tf.constant(['butterflyflybutterfly', 'flyflybutterbutter']))
<tf.RaggedTensor [[1, 2, 2, 1, 2], [2, 2, 1, 1]]>
```

Reply:
This will also be potentially resolved in a future PR as needed